src/torrent.cpp

```diff
@@ -5664,99 +5675,66 @@ namespace {
 	void torrent::on_file_priority(disk_io_job const* j)
 	{
 		dec_refcount("file_priority");
+
 		boost::scoped_ptr<std::vector<boost::uint8_t> > p(j->buffer.priorities);
-		if (m_file_priority == *p) return;
 
-		// in this case, some file priorities failed to get set
-		m_file_priority = *p;
-		update_piece_priorities();
+		if (j->ret == piece_manager::no_error)
+		{
+			TORRENT_ASSERT(!j->error);
+			m_file_priority = *p;
+			update_piece_priorities();
+		}
+		else
+		{
+			// in this case, some file priorities failed to get set
+			TORRENT_ASSERT(j->error);
+			TORRENT_ASSERT(!m_share_mode);
 
-		if (alerts().should_post<file_error_alert>())
-			alerts().emplace_alert<file_error_alert>(j->error.ec
-				, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+			if (alerts().should_post<file_error_alert>())
+				alerts().emplace_alert<file_error_alert>(j->error.ec
+					, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
 
-		set_error(j->error.ec, j->error.file);
-		pause();
+			set_error(j->error.ec, j->error.file);
+			pause();
+		}
 	}
 
 	void torrent::prioritize_files(std::vector<int> const& files)
 	{
 		INVARIANT_CHECK;
 
-		// this call is only valid on torrents with metadata
-		if (!valid_metadata() || is_seed()) return;
```

Comment:
I take it it's a deliberate change to allow changing file priorities even when you're a seed. What's the rationale there? the problem of settings files to priority 0 once they're downloaded is that it may be very expensive to copy into the partfile.

Reply:
This check was relocated to the line 5715.